### PR TITLE
+htc add convenience helpers for language negotiation to `HttpRequest`

### PR DESCRIPTION
**Pull request by sirthias**
Closing a small feature gap.
We already have similar convenience helpers for `MediaType`, `HttpCharset` and `HttpEncoding` and should also provide them for `Language` out of the box.

/cc @jrudolph

### CODE REVIEW

File: akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala [hunk 1]
```diff
@@ -244,6 +254,21 @@ final case class HttpRequest(method: HttpMethod = HttpMethods.GET,
...
+   */
+  def isLanguageAccepted(language: Language, ranges: Seq[LanguageRange] = acceptedLanguageRanges): Boolean =
+    qValueForLanguage(language, ranges) > 0f
+
+  /**
+   * Returns the q-value that the client (implicitly or explicitly) attaches to the given language.
+   */
+  def qValueForLanguage(language: Language, ranges: Seq[LanguageRange] = acceptedLanguageRanges): Float =
+    ranges match {
+      case Nil ⇒ 1.0f // http://tools.ietf.org/html/rfc7231#section-5.3.1
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/scaladsl/model/HttpMessage.scala]: thanks for the reference here

**Comment by ktoso** [comment]
LGTM, thanks!

**Comment by ktoso** [comment]
How about adding a very simple test; for the sake of good coverage of HttpMessage?

**Comment by ktoso** [comment]
Created an issue to track this addition more explicitly (in the milestone) https://github.com/akka/akka/issues/19023

**Comment by sirthias** [comment]
Ok, I'll add some tests.

**Comment by ktoso** [comment]
Thanks! I know it's a small addition, thanks for adding a test or two :-)

**Comment by ktoso** [comment]
Failure was unrelated: TcpSpec https://github.com/akka/akka/issues/19025

**Comment by sirthias** [comment]
Discovered a few more things we can improve around language support.
I'll beef up and complete this PR tomorrow morning...
(Almost done, only a few more tests missing, but I have to run off now.)

**Comment by ktoso** [comment]
Thanks a lot @sirthias! Sounds good, it should make it into the release if it makes it in the first half of the day I think

**Comment by sirthias** [comment]
Ok, this thing turned out to become a little bit bigger than initially expected.
It turned out that our content negotiation logic is currently not particularly modular, so I refactored a few things to make working with languages and language-ranges more convenient.
Also, we now have a `selectPreferredLanguage` directive which encapsulates language negotiation and properly tests the low-level stuff.

I'll refactor the other content negotiation stuff (media-types, charsets, encodings) accordingly, but probably won't get that in before tonight or Monday.

**Comment by ktoso** [comment]
Thanks for investigating further, the plan sounds very good - the rest will then arrive with 2.0-M3, and today we can ship this patch - Reviewing now.

File: akka-http-core/src/main/scala/akka/http/impl/model/parser/AcceptLanguageHeader.scala [hunk 2]
```diff
@@ -23,5 +23,5 @@ private[parser] trait AcceptLanguageHeader { this: Parser with CommonRules with
     }
   }
 
-  def `language-range` = rule { ws('*') ~ push(LanguageRange.`*`) | language }
+  def `language-range` = rule { ws('*') ~ push(LanguageRange.`*`) | language ~> (LanguageRange(_)) }
```
Comment (ktoso) [inline akka-http-core/src/main/scala/akka/http/impl/model/parser/AcceptLanguageHeader.scala]: :+1:

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/MiscDirectives.scala [hunk 3]
```diff
@@ -41,6 +41,29 @@ trait MiscDirectives {
...
+   * the order of the arguments is used as a tie breaker (First one wins).
+   */
+  def selectPreferredLanguage(first: Language, more: Language*): Directive1[Language] = {
+    val available = first :: List(more: _*) // we use List rather than Seq since element count is likely very small
+    BasicDirectives.extractRequest.map { req ⇒
+      val sortedWithQValues = available.zip(available.map(req.qValueForLanguage(_))).sortBy(-_._2)
+      val firstBest = sortedWithQValues.head
+      val moreBest = sortedWithQValues.tail.takeWhile(_._2 == firstBest._2)
+      if (moreBest.nonEmpty) {
+        // we have several languages that have the same qvalue, so we pick the one the `Accept-Header` lists first
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/MiscDirectives.scala]: yep, makes sense

**Comment by ktoso** [comment]
LGTM, will merge after validation, thanks @sirthias!
Looking forward to the follow-up next week then

**Comment by 2beaucoup** [comment]
related: https://github.com/spray/spray/pull/839

**Comment by sirthias** [comment]
Actually, Konrad, I just discovered that I can still improve things here somewhat.
How long before you cut M2?

**Comment by ktoso** [comment]
I'm ok with waiting for this PR, how about you guys @drewhk @johanandren?

**Comment by sirthias** [comment]
Ok, just pushed another update.
Good now from my side (pending validation).

**Comment by drewhk** [comment]
We have still time to merge this. Let's timebox PRs until 14:00?

**Comment by ktoso** [comment]
Sounds good to me :+1:

**Comment by drewhk** [comment]
LGTM

**Comment by johanandren** [comment]
LGTM!

**Pull request merged**